### PR TITLE
Add copy out to the fallback path in SR invocation of composed op

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70656
* #70464
* __->__ #70871
* #70508
* #70463

We had previously handled reusing memory in the optimized kernel execution path, but not yet handled it if we hit the unoptimized fallback.

Differential Revision: [D33458652](https://our.internmc.facebook.com/intern/diff/D33458652)

### CODE REVIEW

File: test/cpp/tensorexpr/test_kernel.cpp [hunk 1]
```diff
@@ -1551,6 +1552,46 @@ TEST_F(Kernel, RunWithAllocatedOutputs) {
...
+  FuseTensorExprs(graph);
+  Code code(graph, "");
+  InterpreterState interpreter{code};
+  std::vector<IValue> stack = {a, b};
+  interpreter.run(stack);
+  at::Tensor out = pop(stack).toTensor();
+  ASSERT_TRUE(at::allclose(out, ref));
+
+  auto inp_1 = at::ones({4, 4}, TensorOptions(kCPU).dtype(at::kFloat));
+  auto inp_ref = at::ones({4, 4}, TensorOptions(kCPU).dtype(at::kFloat));
```
Comment (navahgar) [inline test/cpp/tensorexpr/test_kernel.cpp:1582]: unused

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp [hunk 2]
```diff
@@ -227,9 +263,44 @@ void insertDynamicShapesGuard(
...
+  auto num_ten_inputs = node->inputs().size();
+  return [num_ten_inputs](Stack& stack) {
+    std::vector<IValue> inputs = pop(stack, num_ten_inputs);
+    // uncommon case - first run
+    if (stack.size() == 0) {
+      for (IValue elem : inputs) {
+        push(stack, std::move(elem));
+      }
+    } else {
+      at::ArrayRef<IValue> outputs = last(stack, num_ten_inputs);
```
Comment (navahgar) [inline torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:284]: Why are we getting the last #inputs from the stack here? Shouldn't we getting the last #outputs instead?

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp [hunk 3]
```diff
@@ -227,9 +263,44 @@ void insertDynamicShapesGuard(
...
+      for (size_t i = 0; i < inputs.size(); ++i) {
+        IValue out = outputs[i];
```
Comment (navahgar) [inline torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:286]: Are you expecting to have the same # of outputs as inputs? Why?

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:352]: #inputs == #outputs. See code comment I added inline

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:354]: See code comment

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp (hunk 2 above)
Comment (navahgar) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:352]: Thanks for the comments. This makes sense now :)

**Review by navahgar — APPROVED** [review]
LGTM

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458652).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458652).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33458652).

**Pull request closed without merging**